### PR TITLE
Upstream useSafeTimeout & add tests

### PR DESCRIPTION
**Pull request by emplums**
This PR upstreams the `useSafeTimeouts` hook and adds some simple tests. 

I've also added documentation, including adding a new `Hooks` subsection to the documentation site 🎉 We might want to rethink how we structure the sidebar later but I think this is fine for now.

Related issue: #1013 

### Merge checklist
- [x] Added or updated TypeScript definitions (`index.d.ts`) if necessary
- [x] Added/updated tests
- [x] Added/updated documentation
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [x] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/components/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

File: docs/content/useSafeTimeout.mdx [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+title: useSafeTimeout
+---
+
+`useSafeTimeout` is a utility Hook that allows you to safely call `setTimeout` and `clearTimeout` within a component, ensuring that all timeouts are cleared when the component unmounts.
```
Comment (emplums) [inline docs/content/useSafeTimeout.mdx:5]: Does `utility hook` sound good here for language??

Comment (T-Hugs) [reply docs/content/useSafeTimeout.mdx:5]: Yeah, I think it's fine. I've only made a conscious effort to be precise with classification of _behaviors_ into "generic behaviors" and "component behaviors." This doesn't really fall into the category of behavior though.

File: docs/content/useSafeTimeout.mdx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    const cancelTimeout = () => {
+      safeClearTimeout(timeoutId)
+    }
+
+    return (
+      <>
+        <Button onClick={handleOnClick}>Click me</Button>
+        <Button onClick={cancelTimeout}>Cancel Timeout</Button>
```
Comment (T-Hugs) [inline docs/content/useSafeTimeout.mdx:27]:
```suggestion
        <Button onClick={cancelTimeout}>Cancel timeout</Button>
```

File: index.d.ts [hunk 3]
```diff
@@ -100,6 +100,14 @@ declare module '@primer/components' {
     setOpen: (open: boolean) => void
   }
 
+  type SetTimeout = (handler: TimerHandler, timeout?: number, ...args: any[]) => number
+  type ClearTimeout = (id: number) => void
+
+  export const useSafeTimeout: () => {
+    safeSetTimeout: SetTimeout,
+    safeClearTimeout: ClearTimeout
+  }
```
Comment (T-Hugs) [inline index.d.ts:109]: Can we avoid writing out the types for SetTImeout and ClearTimeout?

```suggestion
  export const useSafeTimeout: () => {
    safeSetTimeout: typeof window.setTimeout,
    safeClearTimeout: typeof window.clearTimeout
  }
```

File: src/hooks/useSafeTimeout.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import {useCallback, useEffect, useRef} from 'react'
+
+type SetTimeout = (handler: TimerHandler, timeout?: number, ...args: any[]) => number
```
Comment (T-Hugs) [inline src/hooks/useSafeTimeout.ts:3]: Same suggestion about using `typeof window.setTimeout` as above.

File: src/hooks/useSafeTimeout.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  }, [])
+
+  const safeClearTimeout = useCallback((id: number) => {
+    clearTimeout(timers.current[id])
+    delete timers.current[id]
+  }, [])
+
+  useEffect(() => {
+    return () => {
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (T-Hugs) [inline src/hooks/useSafeTimeout.ts:27]: Does this eslint disable do anything?

File: src/hooks/useSafeTimeout.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+type SetTimeout = (handler: TimerHandler, timeout?: number, ...args: any[]) => number
+type ClearTimeout = (id: number) => void
+
+/**
+ * Safely call `setTimeout` and `clearTimeout` within a component.
+ *
+ * This hook ensures that all timeouts are cleared when the component unmounts.
+ */
+export default function useSafeTimeout(): {safeSetTimeout: SetTimeout; safeClearTimeout: ClearTimeout} {
+  const timers = useRef<Record<string, number>>({})
```
Comment (T-Hugs) [inline src/hooks/useSafeTimeout.ts:12]: I think using a `Set` here is a bit cleaner.

```
const timers = useRef<Set<number>>(new Set<number>())
// ...
timers.current.add(id)
// ...
clearTimeout(id)
timers.current.delete(id)
```

File: src/__tests__/useSafeTimeout.tsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { renderHook} from '@testing-library/react-hooks'
+import useSafeTimeout from '../hooks/useSafeTimeout'
+
+test('should call callback after time', async () => {
+  const { result, waitFor } = renderHook(() => useSafeTimeout())
```
Comment (T-Hugs) [inline src/__tests__/useSafeTimeout.tsx:5]: Is there a reason for not using fake timers on this test?

File: src/__tests__/useSafeTimeout.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  const mockFunction = jest.fn()
+  result.current.safeSetTimeout(mockFunction, 300)
+  await waitFor(() => expect(mockFunction).toHaveBeenCalled())
+})
+
+
+test('should clear timeouts', async () => {
+  jest.useFakeTimers();
+  const { result} = renderHook(() => useSafeTimeout())
```
Comment (T-Hugs) [inline src/__tests__/useSafeTimeout.tsx:14]:
```suggestion
  const { result } = renderHook(() => useSafeTimeout())
```

File: src/__tests__/useSafeTimeout.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  const mockFunction = jest.fn()
+  let timeoutId = null
+  timeoutId = result.current.safeSetTimeout(mockFunction, 300)
```
Comment (T-Hugs) [inline src/__tests__/useSafeTimeout.tsx:17]:
```suggestion
  const timeoutId = result.current.safeSetTimeout(mockFunction, 300)
```

File: src/hooks/useSafeTimeout.ts (hunk 4 above)
Comment (emplums) [reply src/hooks/useSafeTimeout.ts:3]: I tried that, but there are some required properties on the native type that doesn't make sense to pass in manually to these wrapper functions:

```
Property '__promisify__' is missing in type '(handler: TimerHandler, timeout?: number | undefined, ...args: any[]) => number' but required in type 'typeof setTimeout'.ts(2322)

````

File: src/hooks/useSafeTimeout.ts (hunk 5 above)
Comment (emplums) [reply src/hooks/useSafeTimeout.ts:27]: no 😂 I just copied it over and forgot to remove

File: src/hooks/useSafeTimeout.ts (hunk 6 above)
Comment (emplums) [reply src/hooks/useSafeTimeout.ts:12]: 👍

**Comment by emplums** [comment]
Note: I had to do an extensive refactoring of our tests to get them to work nicely with typescript after adding the `@testing-library/react-hooks` library 😅

File: src/hooks/useSafeTimeout.ts (hunk 4 above)
Comment (T-Hugs) [reply src/hooks/useSafeTimeout.ts:3]: Oh weird! Thanks for checking.

File: src/__tests__/FilterList.tsx [hunk 10]
```diff
@@ -25,7 +25,7 @@ describe('FilterList', () => {
     expect(render(<FilterList />).type).toEqual('ul')
   })
 
-  it('wraps children in <li>', () => {
-    expect(render(<FilterList>Hello</FilterList>).children.pop().type).toEqual('li')
-  })
+  // it.skip('wraps children in <li>', () => {
+  //   expect(render(<FilterList>Hello</FilterList>).children.pop().type)?.toEqual('li')
+  // })
```
Comment (colebemis) [inline src/__tests__/FilterList.tsx:30]: We can just delete this 👍 
```suggestion
  // it.skip('wraps children in <li>', () => {
  //   expect(render(<FilterList>Hello</FilterList>).children.pop().type)?.toEqual('li')
  // })
```

File: src/__tests__/FilterListItem.tsx [hunk 11]
```diff
@@ -5,10 +5,17 @@ import {COMMON} from '../constants'
...
 import {axe, toHaveNoViolations} from 'jest-axe'
 import 'babel-polyfill'
+import { ReactTestRendererJSON } from 'react-test-renderer'
 expect.extend(toHaveNoViolations)
 
+
+function isReactNodeJSON(item: string | ReactTestRendererJSON | undefined): item is ReactTestRendererJSON {
+  return (item as ReactTestRendererJSON).type !== undefined;
+}
+
```
Comment (colebemis) [inline src/__tests__/FilterListItem.tsx:15]: Looks like this isn't being used 

```suggestion

```

File: src/__tests__/useSafeTimeout.tsx [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import { renderHook} from '@testing-library/react-hooks'
```
Comment (colebemis) [inline src/__tests__/useSafeTimeout.tsx:1]: Not sure why prettier didn't catch this:
```suggestion
import { renderHook } from '@testing-library/react-hooks'
```

File: src/utils/testing.tsx [hunk 13]
```diff
@@ -27,16 +41,16 @@ export function mount(component) {
  * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
  * ```
  */
-export function render(component) {
-  return renderer.create(component).toJSON()
+export function render(component: React.ReactElement){
```
Comment (colebemis) [inline src/utils/testing.tsx:44]:
```suggestion
export function render(component: React.ReactElement) {
```

File: src/utils/testing.tsx [hunk 14]
```diff
@@ -177,11 +193,21 @@ export function unloadCSS(path) {
...
+}
+
+interface IBehavesAsComponent {
+  Component: React.FunctionComponent<any>,
+  systemPropArray: any[],
+  toRender?: () => React.ReactElement,
+  options?: IOptions
+}
+
+export function behavesAsComponent({Component, systemPropArray, toRender, options}: IBehavesAsComponent) {
```
Comment (colebemis) [inline src/utils/testing.tsx:210]: We don't usually prefix our interfaces with `I`:
```suggestion
interface Options {
  skipAs?: boolean
  skipSx?: boolean
}

interface BehavesAsComponent {
  Component: React.FunctionComponent<any>,
  systemPropArray: any[],
  toRender?: () => React.ReactElement,
  options?: Options
}

export function behavesAsComponent({Component, systemPropArray, toRender, options}: BehavesAsComponent) {
```

**Review by colebemis — APPROVED** [review]
Nice work! Thanks for dealing with the tests ❤️ Just left a few minor comments.

File: .changeset/spotty-parents-beam.md [hunk 15]
```diff
@@ -0,0 +1,5 @@
+---
+"@primer/components": patch
```
Comment (colebemis) [inline .changeset/spotty-parents-beam.md:2]: I think of `patch` versions as bug fixes, dependency bumps, etc. `useSafeTimeout` seems like a new feature. I'd call it a `minor` release.

```suggestion
"@primer/components": minor
```

**Pull request merged**